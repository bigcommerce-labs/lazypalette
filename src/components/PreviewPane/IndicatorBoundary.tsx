import React, { SFC } from 'react';

import { ErrorIndicator } from '../ErrorIndicator/ErrorIndicator';

interface IndicatorBoundaryProps {
    children: JSX.Element;
    errors: Error[];
    isFetching: boolean;
    clearErrors(): void;
}

export const IndicatorBoundary: SFC = (props: IndicatorBoundaryProps): JSX.Element => {
    if (props.errors.length > 0) {
        return <ErrorIndicator errors={props.errors} clearErrors={props.clearErrors} />;
    } else {
        return props.children;
    }
};
