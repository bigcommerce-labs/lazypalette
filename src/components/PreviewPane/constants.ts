export const VIEWPORT_TYPES = {
    DESKTOP: {
        glyphName: 'desktop',
        viewportHeight: '100%',
        viewportWidth: '100%',
    },
    MOBILE: {
        glyphName: 'mobile',
        viewportHeight: '667px',
        viewportWidth: '375px',
    },
    TABLET: {
        glyphName: 'tablet',
        viewportHeight: '1024px',
        viewportWidth: '768px',
    },
};
