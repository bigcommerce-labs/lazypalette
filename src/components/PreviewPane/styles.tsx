import { theme } from 'pattern-lab';
import styled from 'styled-components';

import { VIEWPORT_TYPES } from './constants';
import { ViewportType } from './PreviewPane';

export const PreviewPaneContainer = styled.div`
    background: ${({ theme: { colors } }) => colors && colors.empty};
    border-radius: 0px;
    box-shadow: ${({ theme: { elevation } }) => elevation && elevation.raised};
    display: flex;
    flex: auto;
    height: calc(100% - 6rem);
    justify-content: center;
    margin: 0.75rem 0.75rem 0 0;
    overflow: hidden;
    width: 0;
`;

interface PreviewPaneIframeProps {
    isFetching: boolean;
    isRotated: boolean;
    viewportType: ViewportType;
}

export const PreviewPaneIframe = styled.iframe.attrs<PreviewPaneIframeProps>({
    height: (props: PreviewPaneIframeProps) =>
        props.isRotated
            ? props.viewportType.viewportWidth
            : props.viewportType.viewportHeight,
    width: (props: PreviewPaneIframeProps) =>
        props.isRotated
            ? props.viewportType.viewportHeight
            : props.viewportType.viewportWidth,
})`
    background-color: #F2F2F2;
    border: 0px;
    ${(props: PreviewPaneIframeProps) => {
        return `
                min-width: ${props.isRotated ? props.viewportType.viewportHeight : props.viewportType.viewportWidth};
                opacity: ${props.isFetching ? 0.5 : 1};
                `;
    }}

    ${(props: PreviewPaneIframeProps) => {
        const borderRadius = props.viewportType === VIEWPORT_TYPES.DESKTOP ? '0px' : '30px';
        let boxShadow;
        let marginBottom;
        let marginTop;
        let paddingBottom;
        let paddingLeft;
        let paddingRight;
        let paddingTop;

        switch (props.viewportType) {
            case VIEWPORT_TYPES.DESKTOP:
                boxShadow = 'none';
                marginBottom = '0px';
                marginTop = '0px';
                paddingBottom = '0px';
                paddingLeft = '0px';
                paddingRight = '0px';
                paddingTop = '0px';
                break;
            case VIEWPORT_TYPES.MOBILE:
                boxShadow = theme.elevation.raised;
                marginBottom = '3rem';
                marginTop = '3rem';
                paddingBottom = '50px';
                paddingLeft = '10px';
                paddingRight = '10px';
                paddingTop = '60px';
                break;
            case VIEWPORT_TYPES.TABLET:
                boxShadow = theme.elevation.raised;
                marginBottom = '3rem';
                marginTop = '3rem';
                paddingBottom = '50px';
                paddingLeft = '20px';
                paddingRight = '20px';
                paddingTop = '65px';
                break;
        }

        if (props.isRotated) {
            [paddingBottom, paddingLeft] = [paddingLeft, paddingBottom];
            [paddingRight, paddingTop] = [paddingTop, paddingRight];
        }

        return `
            border-radius: ${borderRadius};
            box-shadow: ${boxShadow};
            margin-bottom: ${marginBottom};
            margin-top: ${marginTop};
            max-height: calc(100% - ${paddingTop} - ${marginTop});
            padding-bottom: ${paddingBottom};
            padding-left: ${paddingLeft};
            padding-right: ${paddingRight};
            padding-top: ${paddingTop};
        `;
    }}
`;
