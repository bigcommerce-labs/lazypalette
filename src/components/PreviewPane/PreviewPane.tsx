import React, { Component } from 'react';
import { connect, Dispatch } from 'react-redux';
import { bindActionCreators } from 'redux';

import {
    buildIframeUrl,
    previewPaneLoaded,
    previewPaneLoading,
    previewPanePageReloaded,
    updatePage,
    UpdatePageAction,
    UpdatePagePayload
} from '../../actions/previewPane';
import { State } from '../../reducers/reducers';
import ChannelService from '../../services/previewPane/channelService/channelService';

import { VIEWPORT_TYPES } from './constants';
import { PreviewPaneContainer, PreviewPaneIframe } from './styles';

export interface ViewportType {
    glyphName: string;
    viewportHeight: string;
    viewportWidth: string;
}

interface SetCookiesParams {
    variationId: string;
    versionId: string;
    lastCommitId: string;
    configurationId: string;
}

interface PreviewPaneProps {
    versionId: string;
    variationId: string;
    configurationId: string;
    lastCommitId: string;
    fontUrl: string | null;
    isFetching: boolean;
    isRotated: boolean;
    needsForceReload: boolean;
    page: string;
    iframeUrl: string;
    viewportType: ViewportType;
    buildIframeUrl(page: string): void;
    previewPaneLoaded(): void;
    previewPaneLoading(): void;
    previewPanePageReloaded(): void;
    updatePage(payload: UpdatePagePayload): UpdatePageAction;
}

class PreviewPane extends Component<PreviewPaneProps> {
    private channelService: ChannelService;
    private iframeRef: HTMLIFrameElement;

    componentDidUpdate(prevProps: PreviewPaneProps) {
        const { configurationId, fontUrl, lastCommitId, needsForceReload, variationId, versionId } = this.props;

        /**
         * Structural updates can occur when the user makes a settings change which would modify the DOM of the page.
         * If the variation changes, this should be considered a structural update.
         * Example: User changes the number of featured products on the homepage.
         */
        if (needsForceReload || prevProps.variationId !== variationId) {
            this.broadcastForceReload({
                configurationId,
                lastCommitId,
                variationId,
                versionId,
            });

            return;
        }

        /**
         * Optimized / Progressive updates
         * For CSS & Font changes, we can do better than reloading the entire page. For CSS, we can inject a new style
         * tag for the browser to pick up new styling. We can use a similar strategy for fonts, just with font links.
         */
        const previousConfigurationId = prevProps.configurationId;
        if (previousConfigurationId && previousConfigurationId !== configurationId) {
            this.broadcastReloadStylesheets(configurationId);

            if (fontUrl && fontUrl !== prevProps.fontUrl) {
                this.broadcastFontChange(fontUrl);
            }

            return;
        }
    }

    render(): JSX.Element {
        const { isFetching, isRotated, iframeUrl, viewportType } = this.props;

        return (
            <PreviewPaneContainer showBorder={viewportType === VIEWPORT_TYPES.DESKTOP}>
                {iframeUrl &&
                    <PreviewPaneIframe
                        innerRef={(x: HTMLIFrameElement) => (this.iframeRef = x)}
                        isFetching={isFetching}
                        isRotated={isRotated}
                        onLoad={this.onLoad}
                        src={iframeUrl}
                        viewportType={viewportType}
                    />
                }
            </PreviewPaneContainer>
        );
    }

    private broadcastFontChange = (fontUrl: string) => {
        this.props.previewPaneLoading();

        if (!this.channelService) {
            this.raceConditionHandler();

            return;
        }

        this.channelService.safeBroadcast({
            error: (error: any) => { return; }, // TODO
            method: 'add-font',
            params: { fontUrl },
            success: (data: any) => this.props.previewPaneLoaded(),
        }, this.raceConditionHandler);
    };

    private broadcastReloadStylesheets = (configurationId: string) => {
        this.props.previewPaneLoading();

        if (!this.channelService) {
            this.raceConditionHandler();

            return;
        }

        this.channelService.safeBroadcast({
            error: (error: any) => { return; }, // TODO
            method: 'reload-stylesheets',
            params: { configurationId },
            success: (data: any) => this.props.previewPaneLoaded(),
        }, this.raceConditionHandler);
    };

    private broadcastForceReload = ({ variationId, versionId, configurationId, lastCommitId }: SetCookiesParams) => {
        if (!this.channelService) {
            this.raceConditionHandler();

            return;
        }

        this.channelService.safeBroadcast({
            error: (error: any) => { return; }, // TODO
            method: 'set-cookie',
            params: {
                configurationId,
                reloadPage: true,
                // sessionId === lastCommitId
                // See: https://github.com/bigcommerce-labs/stencil-preview-sdk/blob/master/src/app/iframe-sdk.js#L99
                sessionId: lastCommitId,
                variationId,
                versionId,
            },
            success: (data: any) => {
                this.props.previewPanePageReloaded();
            },
        }, this.raceConditionHandler);
    };

    private raceConditionHandler = () => {
        this.props.buildIframeUrl(this.props.page);
        this.props.previewPanePageReloaded();
    };

    private onLoad = () => {
        // TODO: We know the iframe has loaded but we dont know if the page which was loaded supports store design. If
        // there is a link to an external page or we load an error page, we will see the onLoad handler is called, but
        // the stencilPreviewSdk will not be available.

        // Dismiss the overlay when the content of the preview pane loads
        this.props.previewPaneLoaded();

        const iframeWindow = this.iframeRef.contentWindow;
        if (!iframeWindow) {
            throw new Error('iframe contentWindow unavailable');
        }

        // Keep the Redux Store in sync with the users page navigation.
        const page = iframeWindow.location.pathname;
        if (page !== this.props.page) {
            this.props.updatePage({ page });
        }

        // Ensure we have a channelService set up when the iframe loads. Since we need to have the iframe reference
        // available to stand up the channel, we need to make sure this is done once the iframe loads.
        // TODO: Look into other places for this to live. We only need to instantiate the ChannelService once, since
        // onLoad is called each time the user navigates, we need to make sure the ChannelService client is a singleton
        // instance. We can definitely improve this.
        // TODO: Handle SdkReady / SdkNotReady signals
        this.channelService = new ChannelService({
            sdkNotReady: () => { return; },
            sdkReady: () => { return; },
            window: iframeWindow,
        });
    };
}

const mapStateToProps = (state: State): Partial<PreviewPaneProps> => {
    return {
        ...state.previewPane,
        ...{
            configurationId: state.theme.configurationId,
            lastCommitId: state.theme.lastCommitId,
            variationId: state.theme.variationId,
            versionId: state.theme.versionId,
        },
    };
};

const mapDispatchToProps = (dispatch: Dispatch): Partial<PreviewPaneProps> => bindActionCreators({
    buildIframeUrl,
    previewPaneLoaded,
    previewPaneLoading,
    previewPanePageReloaded,
    updatePage,
}, dispatch);

export default connect<Partial<PreviewPaneProps>, Partial<PreviewPaneProps>, {}, State>(
    mapStateToProps, mapDispatchToProps
)(PreviewPane);
