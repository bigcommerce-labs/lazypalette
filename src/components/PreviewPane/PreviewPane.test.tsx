import { shallow } from 'enzyme';
import React from 'react';
import createMockStore from 'redux-mock-store';
import PreviewPane from './PreviewPane';

it('renders', () => {
  const store = createMockStore([])({ previewPane: {} });
  const previewPane = shallow(<PreviewPane />, { context: { store }});

  expect(previewPane).toMatchSnapshot();
});
