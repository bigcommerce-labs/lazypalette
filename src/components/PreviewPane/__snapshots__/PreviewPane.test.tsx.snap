// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewPane render renders the component 1`] = `
<Connect(PreviewPane)>
  <PreviewPane
    buildIframeUrl={[Function]}
    configurationId="26a5d5b0-72ff-0134-79ee-0242ac112a0f"
    fontUrl={null}
    iframeUrl="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
    isFetching={false}
    isRotated={false}
    lastCommitId=""
    needsForceReload={false}
    page="/"
    previewPaneLoaded={[Function]}
    previewPanePageReloaded={[Function]}
    updatePage={[Function]}
    variationId="3fde4290-8d3c-0136-f365-0242ac11000d"
    versionId="26a5d5b0-72ff-0134-79ee-0242ac112a0f"
    viewportType={
      Object {
        "glyphName": "desktop",
        "tooltip": "Desktop preview",
        "viewportHeight": "100%",
        "viewportWidth": "100%",
      }
    }
  >
    <styled.div
      showBorder={true}
      theme={
        Object {
          "colors": Object {
            "empty": "#FFFFFF",
          },
          "elevation": Object {
            "raised": 100,
          },
        }
      }
    >
      <div
        className="sc-bdVaJa kqEwVy"
      >
        <styled.iframe
          innerRef={[Function]}
          isFetching={false}
          isRotated={false}
          onLoad={[Function]}
          src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
          theme={
            Object {
              "colors": Object {
                "empty": "#FFFFFF",
                "stroke": "#CFD6E5",
              },
              "elevation": Object {
                "raised": "0 1px 6px rgba(48, 53, 64, 0.2)",
              },
            }
          }
          viewportType={
            Object {
              "glyphName": "desktop",
              "tooltip": "Desktop preview",
              "viewportHeight": "100%",
              "viewportWidth": "100%",
            }
          }
        >
          <iframe
            className="sc-bwzfXH eWDDUF"
            height="100%"
            onLoad={[Function]}
            src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
            width="100%"
          />
        </styled.iframe>
      </div>
    </styled.div>
  </PreviewPane>
</Connect(PreviewPane)>
`;

exports[`PreviewPane render when the iframeUrl is not ready does not render the inner iframe 1`] = `
<Connect(PreviewPane)>
  <PreviewPane
    buildIframeUrl={[Function]}
    configurationId="26a5d5b0-72ff-0134-79ee-0242ac112a0f"
    fontUrl={null}
    iframeUrl=""
    isFetching={false}
    isRotated={false}
    lastCommitId=""
    needsForceReload={false}
    page="/"
    previewPaneLoaded={[Function]}
    previewPanePageReloaded={[Function]}
    updatePage={[Function]}
    variationId="3fde4290-8d3c-0136-f365-0242ac11000d"
    versionId="26a5d5b0-72ff-0134-79ee-0242ac112a0f"
    viewportType={
      Object {
        "glyphName": "desktop",
        "tooltip": "Desktop preview",
        "viewportHeight": "100%",
        "viewportWidth": "100%",
      }
    }
  >
    <styled.div
      showBorder={true}
      theme={
        Object {
          "colors": Object {
            "empty": "#FFFFFF",
          },
          "elevation": Object {
            "raised": 100,
          },
        }
      }
    >
      <div
        className="sc-bdVaJa kqEwVy"
      />
    </styled.div>
  </PreviewPane>
</Connect(PreviewPane)>
`;

exports[`PreviewPane viewport switching when the viewport type is set to DESKTOP updates the viewportType correctly 1`] = `
<iframe
  class="sc-bwzfXH eWDDUF"
  height="100%"
  src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
  width="100%"
/>
`;

exports[`PreviewPane viewport switching when the viewport type is set to DESKTOP when rotated updates the rotated viewportType correctly 1`] = `
<iframe
  class="sc-bwzfXH eWDDUF"
  height="100%"
  src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
  width="100%"
/>
`;

exports[`PreviewPane viewport switching when the viewport type is set to MOBILE updates the viewportType correctly 1`] = `
<iframe
  class="sc-bwzfXH kpOTJA"
  height="667px"
  src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
  width="375px"
/>
`;

exports[`PreviewPane viewport switching when the viewport type is set to MOBILE when rotated updates the rotated viewportType correctly 1`] = `
<iframe
  class="sc-bwzfXH cxHELo"
  height="375px"
  src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
  width="667px"
/>
`;

exports[`PreviewPane viewport switching when the viewport type is set to TABLET updates the viewportType correctly 1`] = `
<iframe
  class="sc-bwzfXH jHXfir"
  height="1024px"
  src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
  width="768px"
/>
`;

exports[`PreviewPane viewport switching when the viewport type is set to TABLET when rotated updates the rotated viewportType correctly 1`] = `
<iframe
  class="sc-bwzfXH gZhnve"
  height="768px"
  src="/?stencilEditor=26a5d5b0-72ff-0134-79ee-0242ac112a0f@26a5d5b0-72ff-0134-79ee-0242ac112a0f"
  width="1024px"
/>
`;
